tests/core/fused_gromov_wasserstein_test.py

```diff
@@ -0,0 +1,249 @@
+      return gromov_wasserstein.gromov_wasserstein(
+        geom_x, geom_y, geom_xy=geom_xy, fused_penalty=fused_penalty, a=a, b=b, epsilon=1.0, max_iterations=10,
+        sinkhorn_kwargs=sinkhorn_kwargs).reg_gw_cost
+
+    grad_matrices = [None, None]
+    for i, implicit in enumerate([True, False]):
+      reg_gw_and_grad = jax.value_and_grad(reg_gw, argnums=(0, 1,))
+      _, grad_reg_gw = reg_gw_and_grad(self.x, self.y, self.x_2, self.y_2, self.fused_penalty, self.a, self.b, implicit)
+      grad_matrices[i] = grad_reg_gw
+      self.assertIsNot(jnp.any(jnp.isnan(grad_reg_gw[0])), True)
+      self.assertIsNot(jnp.any(jnp.isnan(grad_reg_gw[1])), True)
+    self.assertAllClose(grad_matrices[0][0], grad_matrices[1][0],
+                        rtol=1e-02, atol=1e-02)
+    self.assertAllClose(grad_matrices[0][1], grad_matrices[1][1],
+                        rtol=1e-02, atol=1e-02)
+
+  @parameterized.parameters([True], [False])
+  def test_gradient_fused_gromov_wasserstein_geometry(self, lse_mode):
+    """Test gradient w.r.t. cost matrices."""
+
+    def reg_gw(cx, cy, cxy, fused_penalty, a, b, implicit):
+      geom_x = geometry.Geometry(cost_matrix=cx)
+      geom_y = geometry.Geometry(cost_matrix=cy)
+      geom_xy = geometry.Geometry(cost_matrix=cxy)
+      sinkhorn_kwargs = {'implicit_differentiation': implicit,
+                         'max_iterations': 1001, 'lse_mode': lse_mode}
+      return gromov_wasserstein.gromov_wasserstein(
+        geom_x, geom_y, geom_xy=geom_xy, fused_penalty=fused_penalty, a=a, b=b, epsilon=1.0, max_iterations=10,
+        sinkhorn_kwargs=sinkhorn_kwargs).reg_gw_cost
+
+    grad_matrices = [None, None]
+    for i, implicit in enumerate([True, False]):
+      reg_gw_and_grad = jax.value_and_grad(reg_gw, argnums=(0, 1, 2,))
+      _, grad_reg_gw = reg_gw_and_grad(
+        self.cx, self.cy, self.cxy, self.fused_penalty, self.a, self.b, implicit)
+      grad_matrices[i] = grad_reg_gw
+      self.assertIsNot(jnp.any(jnp.isnan(grad_reg_gw[0])), True)
+      self.assertIsNot(jnp.any(jnp.isnan(grad_reg_gw[1])), True)
+    self.assertAllClose(grad_matrices[0][0], grad_matrices[1][0],
+                        rtol=1e-02, atol=1e-02)
+    self.assertAllClose(grad_matrices[0][1], grad_matrices[1][1],
+                        rtol=1e-02, atol=1e-02)
+    self.assertAllClose(grad_matrices[0][2], grad_matrices[1][2],
+                        rtol=1e-02, atol=1e-02)
+
+  def test_adaptive_threshold_fused(self):
+    """Checking solution is improved with smaller threshold for convergence."""
+    geom_x = pointcloud.PointCloud(self.x, self.x)
+    geom_y = pointcloud.PointCloud(self.y, self.y)
+    geom_xy = pointcloud.PointCloud(self.x_2, self.y_2)
+    # without warm start for calls to sinkhorn
+    def loss_thre(threshold):
+      return gromov_wasserstein.gromov_wasserstein(
+          geom_xx=geom_x, geom_yy=geom_y, geom_xy=geom_xy, fused_penalty=self.fused_penalty_2, a=self.a, b=self.b,
```

Comment:
did this work for the other fused_penalty? is there a good reason to introduce it?

Reply:
It seems like the penalty term is too dominant if we take self.fused_penalty instead of self.fused_penalty_2. We get the same values, i.e. not a strict inequality: loss_thre(.1) == loss_thre(.001) == loss_thre(.00001).

Alternatively, we could reduce the influence of the linear term by the following 
self.x_2 = jax.random.uniform(keys[0], (self.n, d_xy))/100
self.y_2 = jax.random.uniform(keys[1], (self.m, d_xy))/100

Here, we reduce the linear penalty term by using smaller distances in the shared space. This would also pass the test.

Please let me know what you prefer.